### ISSUE TITLE
PodFitsHostPorts fail on same hostPort, different hostIP

### ISSUE THREAD

**Issue by CallMeFoxie** [kubernetes/kubernetes#51950]
/kind bug

Hi
I've posted PR few months ago #39369 - regarding validating against combination of hostPort:protocol:hostIP.
It seems that something similar is back, this time with Predicates.

The issue is in `plugin/pkg/scheduler/algorithm/predicates/predicates.go` in `PodFitsHostPorts:880` - it checks only for ports and not for IPs they listen on.

I will try to get PR for this but not if it will be soon.

Basically: one pod SHOULD be able to listen on:
10.1.1.1:8080
10.1.1.2:8080

right now it gets denied with PodFitsHostPorts error.
It was working back in 1.5 or whichever release was in February this year. It seems to have been broken around 1.7.0 release based on Git Blame.

Ashley

**Comment by WIZARD-CXY** [comment]
@CallMeFoxie if you can not fix it soon, I can give a hand to fix this.

**Comment by WIZARD-CXY** [comment]
After a quick search of the code I think I suspect this pr https://github.com/kubernetes/kubernetes/pull/42524/files causes the regression. The time matches

**Comment by k82cn** [comment]
it seems #39369 only enhance validation instead of pod predicates.

**Comment by WIZARD-CXY** [comment]
I think update the function ```updateUsedPorts``` which brought by https://github.com/kubernetes/kubernetes/pull/42524/files  and change ```usedPorts        map[int]bool```  to map[string]bool, in which key(string) is hostIp/hostport/protocol and add some unit test is enough.WDYT

**Comment by CallMeFoxie** [comment]
That's ~identical to what i did in validation so should be fine i think.

**Comment by WIZARD-CXY** [comment]
@k82cn I look through the code and find out the scheduler only predicates using hostPort, not "hostPort protocol" or "hostip hostPort protocol", I wonder if it is intended, If so, it looks a little surprised to me.

**Comment by k82cn** [comment]
yes, it's the logic since 1.5.

**Comment by WIZARD-CXY** [comment]
Ok then, why is it intended? Do we need to change it? @davidopp @timothysc

**Comment by timothysc** [comment]
It's intended.  IIRC hostports are special, in that they are propagated to all nodes in the cluster and should be treated with extra care.  They were a way of exposing services pre-ingress-controller for folks who didn't have native cloud load-balancer integration.

**Comment by CallMeFoxie** [comment]
hostports? Don't you mean those nodeport services?

**Comment by timothysc** [comment]
sorry... you are correct I transposed those when writing that.

**Comment by WIZARD-CXY** [comment]
I think we should add it. "Hostport and hostIP" can be useful when using a hostNetwork pod and wants to bind different ip on the host just like @CallMeFoxie's user case.Besides adding protocol is necessary too IMHO, for example a dns server may use both 53/udp and 53/tcp.

**Comment by WIZARD-CXY** [comment]
@k82cn @timothysc WDYT. I think it is not a hard thing to implement and if we have agreement on adding this, I can start to implement it.

**Comment by timothysc** [comment]
@WIZARD-CXY feel free to PR and tag the scheduling sig.

**Comment by WIZARD-CXY** [comment]
cool

**Comment by WIZARD-CXY** [comment]
@CallMeFoxie as the pr #52421 has been merged, you can now deploy pod with different ```hostip:hostPort:protocol``` combination now.
